Do not run tests that have `ensure` instead of `skip`

yahonda opened the pull request:
### Summary

This commit enables `Minitest/SkipEnsure` available since RuboCop Minitest 0.20
via https://github.com/rubocop/rubocop-minitest/pull/171

This cop does not support auto-correction, these offenses are fixed
manually.

- Add `begin` .. `end` not to call `ensure` when it is skipped.

- Removed `if file` at ensure in `test_pool_from_any_process_for_uses_most_recent_spec`
because the `file` should have been created inside the test code when
it is not skipped.

### Other Information

This cop idea was based on https://github.com/rails/rails/pull/42691.
One of the tests had raised `ActiveRecord::StatementInvalid` exception
because it attempted to drop a column that was added inside the test code.
Adding column inside the test was skipped but drop column inside the ensure was called.

https://buildkite.com/rails/rails/builds/78986#ccd268e5-7a1f-4c4f-9631-9cb4bb61ff4e/1086-1122



<!-- Provide a general description of the code changes in your pull
request... were there any bugs you had fixed? If so, mention them. If
these bugs have open GitHub issues, be sure to tag them here as well,
to keep the conversation linked together. -->


<!-- If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

If you are updating any of the CHANGELOG files or are asked to update the
CHANGELOG files by reviewers, please add the CHANGELOG entry at the top of the file.

Finally, if your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

Thanks for contributing to Rails! -->

<!--
Note: Please avoid making *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.
Create a pull request when it is ready for review and feedback
from the Rails team :).
-->

yahonda commented:
https://github.com/rails/rails/pull/45216/files?diff=unified&w=1 shows the diff without whitespaces.

koic commented on .rubocop.yml:
```diff
@@ -310,3 +310,6 @@ Performance/OpenStruct:
   Enabled: true
+
+Minitest/SkipEnsure:
+  Enabled: true
```
That’s just my two cents. It seems to be sorted by ASCII, so it would be better to move it above `Minitest/UnreachableAssertion` cop.

yahonda replied:
Good catch. Fixed.

matthewd commented on activerecord/test/cases/connection_adapters/connection_handler_test.rb:
```diff
@@ -406,33 +406,33 @@ def test_retrieve_connection_pool_copies_schema_cache_from_ancestor_pool
 
-          rd.close
-        ensure
-          if file
```
Hmm... as well as disliking that this makes the test more indented, I disagree that `if file` is redundant: if `Tempfile.new` raises, this will have a bad time.

rafaelfranca commented:
A while back (many years ago) I went trough all our skipped tests and changed to not use `skip` and instead conditionally define them, since with skip it is hard to know if the test should run and it is skipped because of some bug, or if the test should not even run. Can't we do that for those tests instead of autocorrect the cop to use `begin/ensure`?

yahonda commented:
Thanks for the review.

> changed to not use skip and instead conditionally define them

Would you tell me some of these changes for reference? because I'd like to know how to rewrite these tests.

rafaelfranca commented:
They look like this
```ruby
unless current_adapter?(:SQLite3Adapter)
  def test_pool_from_any_process_for_uses_most_recent_spec
   # ...
  end
end
```

Here is one example https://github.com/rails/rails/blob/8ef3ddcc81fd9f2cfff340732d0130b0cf82cf95/activerecord/test/cases/adapter_test.rb#L195

yahonda committed:
Do not run tests that have `ensure` instead of `skip`

Even if tests are "skipped" by `Minitest::Assertions#skip`
`ensure` is executed because tests methods are Ruby methods.

Adding conditions outside of the test methods, entire test methods are
executed only when they are necessary.

yahonda commented:
Thanks for the advice. I have updated this pull request as suggested.

yahonda commented:
Let me take a look at CI failures.

yahonda commented on railties/test/console_helpers.rb:
```diff
@@ -18,8 +18,8 @@ def assert_output(expected, io, timeout = 10)
-  def available_pty?
-    defined?(PTY) && PTY.respond_to?(:open)
-  end
+def available_pty?
```
To make available_pty? method available outside of test methods, moving it outside of ConsoleHelpers module as other helper method does.
